### PR TITLE
[SPARK-5801] [core] Avoid creating nested directories.

### PR DESCRIPTION
**Pull request by vanzin**
Cache the value of the local root dirs to use for storing local data,
so that the same directories are reused.

Also, to avoid an extra level of nesting, use a different env variable
to propagate the local dirs from the Worker to the executors. And make
the executor directory use a different name.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala [hunk 1]
```diff
@@ -135,7 +135,7 @@ private[spark] class ExecutorRunner(
       logInfo("Launch command: " + command.mkString("\"", "\" \"", "\""))
 
       builder.directory(executorDir)
-      builder.environment.put("SPARK_LOCAL_DIRS", appLocalDirs.mkString(","))
+      builder.environment.put("SPARK_EXECUTOR_DIRS", appLocalDirs.mkString(File.pathSeparator))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala]: I think these are dumb questions, but why does the fix entail setting a different env variable? and should dirs be joined with a path separator?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala]: Without this, `Utils.scala` would always create a subdirectory for the executor under the directory already created by the Worker. So you'd always have "spark-xxxxx/spark-yyyyy" for every executor. It would be always like that (no extra nesting), but, since I'm changing this, sounded like a good enhancement.

And dirs should be joined with a path separator because that's the right way to do it.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala]: Oh right, this is `path.separator` not `file.separator`, `:` not `/`. 

Yes that's exactly the fix that needs to be made.

`SPARK_EXECUTOR_DIRS` is a new system property right? where is it read? I am missing something.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -679,12 +680,25 @@ private[spark] object Utils extends Logging {
    * If no directories could be created, this will return an empty list.
    */
   private[spark] def getOrCreateLocalRootDirs(conf: SparkConf): Array[String] = {
+    if (localRootDirs == null) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Can this whole mechanism be replaced with a `lazy val` for `localRootDirs`?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: The problem with the lazy val is that you don't get access to the `SparkConf` argument.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: (On a side note, this patch might do some weird things if the method is called with different confs.)

**Comment by vanzin** [comment]
Hold off a little bit on this; let me see how to handle the SparkConf issue (I think Josh added that for the unit tests).

**Comment by andrewor14** [comment]
retest this please. We really need to fix the `JavaAPISuite` at some point.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -728,6 +746,11 @@ private[spark] object Utils extends Logging {
     localDirs
   }
 
+  /** Used by unit tests. Do not call from other places. */
+  private[spark] def clearLocalRootDirs(): Unit = {
```
Comment (Sephiroth-Lin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: May be we can call this function to delete the local root directory in non-yarn mode when application is exited or SparkContext is stoped.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala [hunk 4]
```diff
@@ -345,11 +345,11 @@ private[spark] class Worker(
           }
 
           // Create local dirs for the executor. These are passed to the executor via the
-          // SPARK_LOCAL_DIRS environment variable, and deleted by the Worker when the
+          // SPARK_EXECUTOR_DIRS environment variable, and deleted by the Worker when the
           // application finishes.
           val appLocalDirs = appDirectories.get(appId).getOrElse {
             Utils.getOrCreateLocalRootDirs(conf).map { dir =>
-              Utils.createDirectory(dir).getAbsolutePath()
+              Utils.createDirectory(dir, namePrefix = "executor").getAbsolutePath()
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala]: Do we need to register these for cleanup at shutdown? cf. https://github.com/apache/spark/pull/4759#issuecomment-75937545

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala]: Not needed (see SPARK-4834). That's done by the `maybeCleanupApplication` in this file when the application finishes.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 3 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Note this method doesn't delete anything. There's a separate PR to clean up the root directories (#4759).

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 5]
```diff
@@ -63,6 +63,7 @@ private[spark] object Utils extends Logging {
   val random = new Random()
 
   private val MAX_DIR_CREATION_ATTEMPTS: Int = 10
+  @volatile private var localRootDirs: Array[String] = null
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: OK at first I was uneasy about adding more global state to a common `Utils` class, but this reflects configuration that's already global to the whole JVM, coming by sys properties like `SPARK_LOCAL_DIRS`.  Maybe this could be declared nearer its usage but that alone isn't worth changing this. I also don't see a way to avoid the double-checked locking; I couldn't get something clean with `AtomicReference` to work. I think the extra `SPARK_EXECUTOR_DIRS` is worth it; it's not exposed otherwise.

In other words, LGTM. Any other reviews? I'll merge after waiting a short while for comments.
